### ISSUE TITLE
DurableFunction throwing ExecutionEngineException

### ISSUE THREAD

**Issue by eddynaka** [Azure/azure-functions-durable-extension#1965]
### Description
Previously I was using simple AzureFunctions developed in C# (netcoreapp3.1).
Since my functions where timing out because my process might take longer than 10min, I switched to DurableFunction style.

### Expected behavior

Should not throw any exception.

### Actual behavior

From time to time, I can see the following exception:
```
Managed Exception = System.ExecutionEngineException:
CallStack - Managed Exception
========================================================
CallStack - Crashing Thread
========================================================
     GCFrame
     HelperMethodFrame_1OBJ
     DurableTask.AzureStorage.TimeoutHandler+<ExecuteWithTimeout>d__1`1[[System.__Canon, System.Private.CoreLib]].MoveNext()
     System.Threading.ExecutionContext.RunInternal(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object)
     System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1+AsyncStateMachineBox`1[[System.__Canon, System.Private.CoreLib],[DurableTask.AzureStorage.TimeoutHandler+<ExecuteWithTimeout>d__1`1[[System.__Canon, System.Private.CoreLib]], DurableTask.AzureStorage]].MoveNext(System.Threading.Thread)
     System.Threading.Tasks.AwaitTaskContinuation.RunOrScheduleAction(System.Runtime.CompilerServices.IAsyncStateMachineBox, Boolean)
     System.Threading.Tasks.Task.RunContinuations(System.Object)
     System.Threading.Tasks.Task.FinishContinuations()
     System.Threading.Tasks.Task.TrySetResult()
     System.Threading.Tasks.Task+DelayPromise.CompleteTimedOut()
     System.Threading.Tasks.Task+DelayPromise+<>c.<.ctor>b__1_0(System.Object)
     System.Threading.TimerQueueTimer.CallCallback(Boolean)
     System.Threading.TimerQueueTimer.Fire(Boolean)
     System.Threading.TimerQueueTimer.System.Threading.IThreadPoolWorkItem.Execute()
     System.Threading.ThreadPoolWorkQueue.Dispatch()
     System.Threading._ThreadPoolWaitCallback.PerformWaitCallback()
     DebuggerU2MCatchHandlerFrame

 Native Call Stack
========================================================
ntdll!NtTerminateProcess
ntdll!RtlExitUserProcess
kernel32!ExitProcessImplementation
coreclr!SafeExitProcess
coreclr!EEPolicy::HandleFatalError
coreclr!SystemNative::GenericFailFast
coreclr!SystemNative::FailFast
```

### Relevant source code snippets
```csharp
// insert code snippet here
```

### Known workarounds
> Provide a description of any known workarounds you used.

### App Details

- **Durable Functions extension version (e.g. v1.8.3)**:
- **Azure Functions runtime version (1.0 or 2.0)**:
- **Programming language used**:

### Screenshots
> If applicable, add screenshots to help explain your problem.

### If deployed to Azure
> We have access to a lot of telemetry that can help with investigations. Please provide as much of the following information as you can to help us investigate!

- **Timeframe issue observed**:
- **Function App name**:
- **Function name(s)**:
- **Azure region**:
- **Orchestration instance ID(s)**:
- **Azure storage account name**:

> If you don't want to share your Function App or storage account name GitHub, please at least share the orchestration instance ID. Otherwise it's extremely difficult to look up information.

**Comment by ConnorMcMahon** [comment]
This is unfortunately occasionally expected behavior currently. 

v9.3.1 of the Storage SDK that we use internally occasionally deadlocks in a way that causes the storage request stalls out. We wrap all of our storage requests in `DurableTask.AzureStorage.TimeoutHandler.ExeucuteWithTimeout` to catch this. The first few times we hit this, we will just retry the operation, but once we hit this 5 times, we kill the process, as we are afraid of leaking too many threads.

The long-term fix is for us to migrate to a later version of the SDK, which we have plans to upgrade to the v12 storage SDKs in the near future. Unfortunately, adding a new dependency is technically a breaking change, so we will have to navigate that upgrade carefully, possibly with a new major version of Durable Functions.

**Comment by eddynaka** [comment]
Hi @ConnorMcMahon ,

is there a way to minimize the issue?
Is it related to the number of requests? the time of it? what are the variables that can cause this?

Thank you!

**Comment by ConnorMcMahon** [comment]
It seems tied to network blips during the operation, so the higher the latency of the requests, the more likely you will hit it.

The most common factors impacting this is:
1. CPU usage
2. The size of the storage messages passed around. If you have large inputs/outputs, we generally have to fetch the entire message every replay of the orchestration, so passing around references to large data instead of just returning them directly can greatly reduce the latency of storage requests.
3. Number of concurrent storage requests

**Comment by eddynaka** [comment]
Hi @ConnorMcMahon ,

got it.

(1) my function uses a LOT of CPU because it runs lots and lots of regular expressions + multithreading.
(2) the size of the message is small. Probably a few KB
(3) how can I verify that?

Thanks again.

**Comment by jdluzen** [comment]
This was also happening to me. In my case, there were a _ton_ of network calls from user code all at once in parallel, likely causing socket exhaustion.

**Comment by BenjaminDuke** [comment]
Just an update -> I upped my app service plan to cope with peak demand and haven't had the problem since.

**Comment by davidmrdavid** [comment]
Closing as duplicate of: https://github.com/Azure/durabletask/issues/516
